### PR TITLE
User can now set an instance to be managed by the client

### LINKED ISSUE #43: Event handler leak
There is a kind of "memory leak" in the list of event handlers (the ```_instanceListeners``` attribute of the client object): if event handlers are registered on a channel, they are kept in memory event when the channel no longer exists.

Of course, this can be avoided by explicitly calling the method ```removeAllListeners``` with the previously used events just before the channel disappears, but this behaviour is really unexpected.

If there is no general way to solve this issue, I think it should at least be documented.

### PR DESCRIPTION
**Pull request by samuelg**
A user can now use instance.manageInstance() to let the client manage event listeners' life cycle for the given instance to automatically cleanup event listeners once an instance has gone away.

This allows cleaning up listeners for Channels, Bridges, LiveRecordings, and Playbacks and must be opted in. The behavior of existing applications will remain as is.

This fixes #43

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2426,19 +2426,82 @@ ari.sounds.list()
...
+
+channel.on('ChannelDtmfReceived', handler);
+
+// at some point in your application, remove the event listener to ensure the handler function can be garbage collected
+channel.removeListener('ChannelDtmfReceived', handler);
+// or remove all event listeners for a particular event type
+channel.removeAllListeners('ChannelDtmfReceived');
+```
+
+However, certain resource instance events are unregistered for you after a final event has been received for that resource to minimize possible memory leaks.
```
Comment (chadxz) [inline README.md]: > certain resource instance events are unregistered for you

can you list the ones that are automatically unregistered?

**Comment by chadxz** [comment]
I didn't see a test for the scenario where you have a handler for something like the 'ChannelConnectedLine' event, and then a 'ChannelDestroyed' event is received. Does that scenario automatically remove the 'ChannelConnectedLine' event handler or do you still have to do that manually?

**Comment by samuelg** [comment]
For resources that emit an event that signals the resource went away (Channels, Bridges, Playbacks, and LiveRecordings), all event listeners tied to a specific instance of that resource will be automatically removed since there's no way that those events could occur in the future.

For example, if I had:

```js
channel.on('ChannelDtmfReceived', handler);
channel.on('ChannelConnectedLine', handler2);
```

and then a 'ChannelDestroyed' event came in for that given channel, both the handler for 'ChannelDtmfReceived' and 'ChannelConnectedLine' would be removed automatically.

File: README.md (hunk 1 above)
Comment (samuelg) [inline README.md]: Good idea. Currently only resources of type Channel, Bridge, Playback, and LiveRecording can be automatically cleaned up as they are the only resource that signal an end event.

Comment (matt-jordan) [inline README.md]: You'll run into a problem here with Channels.

Most channels in a Stasis application will never see the `ChannelDestroyed` event - the subscription to the channel is created when the channel enters `Stasis`, and is removed when the channel leaves `Stasis`. As such, the `ChannelDestroyed` event will never be seen by the library, and the subscriptions will persist. The only way you'll see the event is if someone made an explicit subscription to the channel, subscribed to an Endpoint, or subscribed to everything.

I can see two ways of handling this:
1. Maintain a subscription count yourself in the library to channels. A `StasisStart` or `ChannelCreated` bumps the subscription, as does any explicit subscription that gets made. A `StasisEnd` or `ChannelDestroyed` drops the subscription, as does the destruction of an explicit subscription. When the subscription count is 0, purge any listeners.
2. Have the caller determine how they want the lifetime of their listeners to end - on `StasisEnd` or `ChannelDestroyed`. This pushes the event lifetime management onto the caller - which is more work for them - but means you're out of the bookkeeping business.

**Comment by samuelg** [comment]
After testing this out I believe the approach taken will not work well for most real world scenarios. I will open a new PR with more documentation in the README concerning the need to remove event listeners to ensure memory leaks are not a concern but at this point I have no plan to merge code that attempts to auto unregister event listeners.

**Pull request closed without merging**